File: packages/flutter/android/src/main/kotlin/dev/zot/flutter/ZotNativePlugin.kt
package dev.zot.flutter

import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import zot.Session
import zot.Stream
import zot.Zot

class ZotNativePlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
  private val sessions = ConcurrentHashMap<String, Session>()
  private val executor = Executors.newCachedThreadPool()
  private val main = Handler(Looper.getMainLooper())
  private var events: EventChannel.EventSink? = null
  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    MethodChannel(binding.binaryMessenger, "dev.zot/methods").setMethodCallHandler(this)
    EventChannel(binding.binaryMessenger, "dev.zot/events").setStreamHandler(this)
  }
  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) { sessions.values.forEach { it.abort() }; sessions.clear() }
  override fun onListen(arguments: Any?, sink: EventChannel.EventSink) { events = sink }
  override fun onCancel(arguments: Any?) { events = null }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    try {
      val args = call.arguments as? Map<*, *> ?: emptyMap<String, Any>()
      when (call.method) {
        "createSession" -> {
          val provider = args["provider"] as? String ?: "anthropic"
          val session = if (args["accessToken"] != null)
            Zot.newSessionWithOAuth(provider, args["accessToken"] as String, args["accountId"] as? String ?: "", args["model"] as? String ?: "", args["systemPrompt"] as? String ?: "")
          else Zot.newSession(provider, args["apiKey"] as? String ?: "", args["model"] as? String ?: "", args["systemPrompt"] as? String ?: "")
          val id = UUID.randomUUID().toString(); sessions[id] = session; result.success(id)
        }
        "prompt" -> {
          val id = args["sessionId"] as String
          val session = sessions[id] ?: return result.error("zot_session", "Unknown session", null)
          result.success(null)
          executor.execute { session.prompt(args["message"] as String, FlutterStream(id)) }
        }
        "abort" -> { sessions[args["sessionId"]]?.abort(); result.success(null) }
        "closeSession" -> { sessions.remove(args["sessionId"])?.abort(); result.success(null) }
        "exportHistory" -> {
          val session = sessions[args["sessionId"]] ?: return result.error("zot_session", "Unknown session", null)
          executor.execute { main.post { result.success(session.exportHistory()) } }
        }
        "importHistory" -> {
          val session = sessions[args["sessionId"]] ?: return result.error("zot_session", "Unknown session", null)
          executor.execute {
            try { session.importHistory(args["history"] as String); main.post { result.success(null) } }
            catch (error: Exception) { main.post { result.error("zot_history", error.message, null) } }
          }
        }
        "extractOpenAIAccountId" -> result.success(Zot.extractOpenAIAccountID(args["idToken"] as String))
        else -> result.notImplemented()
      }
    } catch (error: Exception) { result.error("zot_error", error.message, null) }
  }

  inner class FlutterStream(private val id: String) : Stream {
    private fun send(event: Map<String, Any>) { main.post { events?.success(mapOf("sessionId" to id, "event" to event)) } }
    override fun onText(delta: String) = send(mapOf("type" to "text", "delta" to delta))
    override fun onEvent(kind: String, payload: String) = send(mapOf("type" to kind, "payload" to payload))
    override fun onError(message: String) = send(mapOf("type" to "error", "message" to message))
    override fun onDone() = send(mapOf("type" to "done"))
  }
}

File: packages/flutter/ios/zot_native/Sources/zot_native/ZotNativePlugin.swift
import Flutter
import UIKit
import Zot

public final class ZotNativePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var sessions: [String: ZotSession] = [:]
    private var sink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let plugin = ZotNativePlugin()
        registrar.addMethodCallDelegate(plugin, channel: FlutterMethodChannel(name: "dev.zot/methods", binaryMessenger: registrar.messenger()))
        FlutterEventChannel(name: "dev.zot/events", binaryMessenger: registrar.messenger()).setStreamHandler(plugin)
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? { sink = events; return nil }
    public func onCancel(withArguments arguments: Any?) -> FlutterError? { sink = nil; return nil }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "createSession":
            var error: NSError?
            let provider = args["provider"] as? String ?? "anthropic"
            let session: ZotSession?
            if let token = args["accessToken"] as? String {
                session = ZotNewSessionWithOAuth(provider, token, args["accountId"] as? String ?? "", args["model"] as? String ?? "", args["systemPrompt"] as? String ?? "", &error)
            } else {
                session = ZotNewSession(provider, args["apiKey"] as? String ?? "", args["model"] as? String ?? "", args["systemPrompt"] as? String ?? "", &error)
            }
            guard let session else { result(FlutterError(code: "zot_session", message: error?.localizedDescription, details: nil)); return }
            let id = UUID().uuidString; sessions[id] = session; result(id)
        case "prompt":
            let id = args["sessionId"] as! String
            guard let session = sessions[id] else { result(FlutterError(code: "zot_session", message: "Unknown session", details: nil)); return }
            result(nil)
            DispatchQueue.global(qos: .userInitiated).async { session.prompt(args["message"] as! String, stream: FlutterZotStream(id: id, sink: self.sink)) }
        case "abort": sessions[args["sessionId"] as! String]?.abort(); result(nil)
        case "closeSession": sessions.removeValue(forKey: args["sessionId"] as! String)?.abort(); result(nil)
        case "exportHistory":
            guard let session = sessions[args["sessionId"] as! String] else { result(FlutterError(code: "zot_session", message: "Unknown session", details: nil)); return }
            DispatchQueue.global(qos: .userInitiated).async { let history = session.exportHistory(); DispatchQueue.main.async { result(history) } }
        case "importHistory":
            guard let session = sessions[args["sessionId"] as! String] else { result(FlutterError(code: "zot_session", message: "Unknown session", details: nil)); return }
            DispatchQueue.global(qos: .userInitiated).async {
                do { try session.importHistory(args["history"] as! String); DispatchQueue.main.async { result(nil) } }
                catch { DispatchQueue.main.async { result(FlutterError(code: "zot_history", message: error.localizedDescription, details: nil)) } }
            }
        case "extractOpenAIAccountId": result(ZotExtractOpenAIAccountID(args["idToken"] as! String))
        default: result(FlutterMethodNotImplemented)
        }
    }
}

private final class FlutterZotStream: NSObject, ZotStreamProtocol {
    let id: String; let sink: FlutterEventSink?
    init(id: String, sink: FlutterEventSink?) { self.id = id; self.sink = sink }
    private func send(_ event: [String: Any]) { DispatchQueue.main.async { self.sink?(["sessionId": self.id, "event": event]) } }
    func onText(_ delta: String?) { send(["type": "text", "delta": delta ?? ""]) }
    func onEvent(_ kind: String?, payload: String?) { send(["type": kind ?? "event", "payload": payload ?? "{}"]) }
    func onError(_ message: String?) { send(["type": "error", "message": message ?? "Unknown error"]) }
    func onDone() { send(["type": "done"]) }
}
